extmod/modbluetooth.c

```diff
@@ -1147,47 +1147,46 @@ mp_bluetooth_gatts_db_entry_t *mp_bluetooth_gatts_db_lookup(mp_gatts_db_t db, ui
 }
 
 int mp_bluetooth_gatts_db_read(mp_gatts_db_t db, uint16_t handle, uint8_t **value, size_t *value_len) {
+    MICROPY_PY_BLUETOOTH_ENTER
     mp_bluetooth_gatts_db_entry_t *entry = mp_bluetooth_gatts_db_lookup(db, handle);
-    if (!entry) {
-        return MP_EINVAL;
-    }
-
-    *value = entry->data;
-    *value_len = entry->data_len;
-    if (entry->append) {
-        entry->data_len = 0;
+    if (entry) {
+        *value = entry->data;
+        *value_len = entry->data_len;
+        if (entry->append) {
+            entry->data_len = 0;
+        }
     }
-
-    return 0;
+    MICROPY_PY_BLUETOOTH_EXIT
+    return entry ? 0 : MP_EINVAL;
 }
 
 int mp_bluetooth_gatts_db_write(mp_gatts_db_t db, uint16_t handle, const uint8_t *value, size_t value_len) {
+    MICROPY_PY_BLUETOOTH_ENTER
     mp_bluetooth_gatts_db_entry_t *entry = mp_bluetooth_gatts_db_lookup(db, handle);
-    if (!entry) {
-        return MP_EINVAL;
-    }
+    if (entry) {
+        if (value_len > entry->data_alloc) {
+            entry->data = m_new(uint8_t, value_len);
```

Comment:
This may raise an exception which would leave the function without releasing the bluetooth lock (yes this would be much clearer without NLR...).  I think it can be fixed pretty easily by using `m_new_maybe()` and then explicitly raising an OOM error after releasing the lock.  Or just returning `MP_ENOMEM`.  But the latter requires all callers to check the return value if they can't guarantee the write won't need to allocate.

Also, related, do we need to check that all calls of `mp_bluetooth_gatts_db_write()` do not have the bluetooth lock?

Reply:
> Also, related, do we need to check that all calls of `mp_bluetooth_gatts_db_write()` do not have the bluetooth lock?

I checked this.  It looks ok, the bluetooth lock is only held in sections of code that are "low level" or call into BTstack.  Ie there are no calls into the uPy runtime (that could raise).

Reply:
There are very few callers of these methods, and most already checked the return value. I've made all uses do so.